addons/content-index/src/main/java/org/commonjava/indy/content/mask/MaskContentManagerDecorator.java

```diff
@@ -0,0 +1,150 @@
+import org.commonjava.maven.galley.model.SpecialPathInfo;
+import org.commonjava.maven.galley.model.Transfer;
+import org.commonjava.maven.galley.model.TransferOperation;
+import org.commonjava.maven.galley.spi.io.SpecialPathManager;
+import org.commonjava.maven.galley.spi.nfc.NotFoundCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.decorator.Decorator;
+import javax.decorator.Delegate;
+import javax.enterprise.inject.Any;
+import javax.inject.Inject;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link ContentManager} decorator that watches the retrieve() methods.
+ * If request targets a Group, repeat the following for each member of the group until the content is found.
+ * If the request targets a RemoteRepository or HostedRepository, check if pathMaskPatterns is non-empty:
+ * <ol>
+ *    <li>if null or empty, proceed normally with retrieve(..) (call the delegate instance)<li/>
+ *    <li>otherwise, iterate through pathMaskPatterns, checking against the requested path:</li>
+ *       <ol>
+ *          <li>If path.startsWith(pattern), return delegate.retrieve(..)</li>
+ *          <li>If path.matches(pattern), return delegate.retrieve(..)</li>
+ *          <li>If no pattern matches, return an "empty" result (null Transfer?)</li>
+ *       </ol>
+ * </ol>
+ *
+ * Created by ruhan on 10/20/16.
+ */
+@Decorator
+public abstract class MaskContentManagerDecorator
+        implements ContentManager
+{
+    final static Logger logger = LoggerFactory.getLogger( MaskContentManagerDecorator.class );
+
+    @Inject
+    private StoreDataManager storeDataManager;
+
+    @Delegate
+    @Any
+    @Inject
+    private ContentManager delegate;
+
+    protected MaskContentManagerDecorator()
+    {
+    }
+
+    protected MaskContentManagerDecorator(final ContentManager delegate, final StoreDataManager storeDataManager)
+    {
+        this.delegate = delegate;
+        this.storeDataManager = storeDataManager;
+    }
+
+    @Override
+    public Transfer retrieve( final ArtifactStore store, final String path )
+            throws IndyWorkflowException
+    {
+        return retrieve( store, path, new EventMetadata() );
+    }
+
+    @Override
+    public Transfer retrieve( final ArtifactStore store, final String path, final EventMetadata eventMetadata )
+            throws IndyWorkflowException
+    {
+        logger.debug( "Checking mask for: {} in: {} type: {}", path, store.getName(), store.getKey().getType() );
+
+        if ( StoreType.group == store.getKey().getType() ) {
+            Group group = (Group) store;
+            for ( StoreKey sk : group.getConstituents()) // TODO: can group contain other groups?
+            {
+                try {
+                    ArtifactStore st = storeDataManager.getArtifactStore(sk);
+                    AbstractRepository repo = (AbstractRepository)st; // TODO: anything not this type?
```

Comment:
Both RemoteRepository and HostedRepository inherit from this abstract class. Group does not. But if you use getOrderedConcreteStoresInGroup(..) then you won't have to worry about detecting the presence of Groups...they will be replaced by their members in the list. You shouldn't have to worry about anything but remote and hosted repos.